### ISSUE TITLE
Advanced BF for restricted models

### ISSUE THREAD

**Issue by mattansb** [easystats/bayestestR#177]
## Interval Null tests
I think it might be possible to extend `bayesfactor_savagedickey` to test against an interval null (kind of a hybrid with ROPE?) according to [Morey and Rouder (2011)](http://sci-hub.tw/http://doi.org/10.1037/a0024377) by extending the order restrictions in `bayesfactor_savagedickey` (which currently allows the one-sided testing).

One sided tests are:
BF<sub>r0</sub> = BF<sub>r1</sub> × BF<sub>10</sub>
<sup>r = the side restriction</sub>

Null interval tests are:
BF<sub>AN</sub> = BF<sub>A1</sub> / BF<sub>N1</sub>
<sup>A = restrict to outside the interval; N = restrict to inside the interval</sub>

This would require users to specify the interval... Maybe change the `hypothesis` to `null` and allow to be of length 1 (for point null) **or** 2 for (interval null)? (All of this is independent of whether or not the test is one/two sided).

Interestingly, if the null interval is in the form of `[Inf,0], then BF = (pd)/(1-pd)...

## Tests parameter restrictions
In essence, we can feed samples of several parameters together and test their joint restrictions. For example, I can have samples for 4 parameters (X1...X4) and we can tests a restricted model of (X1 > 0) & ( X2 > X3 > X4). This will give a BF of the restricted model against the full model (not the null!): BF<sub>r1</sub>.

I've already implemented something like this in [`BFEffects::restrict`](https://github.com/mattansb/BFEffects#using-restrict), but it is heavily dependent on `purrr` and `dplyr`, and only worked with `BayesFactor` objects.

<br><br><br><br><br>
If we agree this is something worth investing in, I think the null-interval would happen for at least v3, and the parameter restrictions would happen maybe around v10 😅...

---
TODO:
- [X] write restrict function
- [X] write interval function
- [x] Choose names / aliases
  - [x] include `rope_range` as default in interval?
- [x] Printing method must indicate a interval vs sdBF
- [x] Plotting method for interval
- [x] docs for all functions!
  - [x] Review
- [x] BF vignette ([outline](https://github.com/easystats/bayestestR/issues/177#issuecomment-505859321))
  - [x] Review
- [x] BF-restricted to the general `bayesfactor` function
- [x] BF-restricted testing
- [x] Update `NEWS.md`
- [x] Update `README`

**Comment by strengejacke** [comment]
Sounds good! I would suggest to open a new branch for this, because we have to re-submit bayestestR in a **very few days**, so I'm not sure if we can implement this in time for the very next update.

**Comment by mattansb** [comment]
Not to worry - I definitely won't start on this for the next 2 weeks or so. But opening a new branch sounds like an excellent idea, as I expect this to be a big shift in the BF functions / docs.

**Comment by mattansb** [comment]
As a stater...

``` r
library(bayestestR)
library(rstanarm)
fit_stan <- stan_glm(mpg ~ wt + cyl + am,
                     data = mtcars)
hyps <- c("am > 0 & cyl < 0",
          "cyl < 0",
          "wt - cyl > 0")

x <- bayesfactor_restrict(fit_stan, hypothesis = hyps)
x
#> Computation of Bayes factors: sampling priors, please wait...
#> # Bayes Factor (Order-Restriction)
#> 
#>        Hypothesis P(Prior) P(Posterior) Bayes Factor
#>  am > 0 & cyl < 0     0.25        0.534         2.16
#>           cyl < 0     0.50        1.000         2.01
#>      wt - cyl > 0     0.52        0.088         0.17
#> ---
#> Bayes factors for the restricted movel vs. the un-restricted model.

colnames(x)
#> [1] "Hypothesis"     "Prior_prob"     "Posterior_prob" "BF"
```

<sup>Created on 2019-06-21 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by mattansb** [comment]
It seems that for factors, order restrictions only give the desired results when using `contr.bayes` contrasts (from [bfrms](https://github.com/bayesstuff/bfrms))...

**Comment by mattansb** [comment]
``` r
library(bayestestR)
prior <- data.frame(X = rnorm(4000))
posterior <- data.frame(X = rnorm(4000,.5,.4))

# point null
(x <- bayesfactor_savagedickey(posterior,prior, null = 0)) # null has replaced hypothesis
#> Loading required namespace: logspline
#> # Bayes Factor (Savage-Dickey density ratio)
#> 
#>  Parameter Bayes Factor
#>          X         0.83
#> ---
#> Evidence Against The Null: [ 0 ]

# interval null
(x <- bayesfactor_savagedickey(posterior,prior, null = c(-0.1, 0.1))) # results are very close for such a small interval
#> # Bayes Factor (Savage-Dickey density ratio)
#> 
#>  Parameter Bayes Factor
#>          X         0.82
#> ---
#> Evidence Against The Null: [ -0.1, 0.1 ]

plot(x) # this still needs work, but technically works
#> Registered S3 methods overwritten by 'ggplot2':
#>   method         from 
#>   [.quosures     rlang
#>   c.quosures     rlang
#>   print.quosures rlang
```

![](https://i.imgur.com/hjEZU1C.png)

``` r

(x <- bayesfactor_restrict(posterior,prior, hypothesis = c("X > .1 | X < -.1","-.1 < X & X < .1")))
#> # Bayes Factor (Order-Restriction)
#> 
#>        Hypothesis P(Prior) P(Posterior) Bayes Factor
#>  X > .1 | X < -.1    0.912        0.903         0.99
#>  -.1 < X & X < .1    0.088        0.097          1.1
#> ---
#> Bayes factors for the restricted movel vs. the un-restricted model.

x$BF[1] / x$BF[2] # this is in theory the same as the interval null, but computed differently.
#> [1] 0.8980288
# nice to see that it (somewhat) converges (sanity check)
```

<sup>Created on 2019-06-22 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>


Technically the interval-null is a probability ratio, and not a savage dickey density ratio... But it feels like these belong together, as their function purpose are the same - compute BFs based on prior and posterior distributions (**not** marginal likelihoods)... but so does `bayesfactor_restrict`, and it does get a separate function because the "what" the is being tested is different (but as you see above, in some cases not too different)...
Any ideas how to deal with all of this? Simply specify in the docs? Make a separate function for the interval test? Make an aliased function/s?

**Comment by DominiqueMakowski** [comment]
That looks great though!



> Technically the interval-null is a probability ratio, and not a savage dickey density ratio... But it feels like these belong together, as their function purpose are the same

I tend to agree. I think it's a good idea to regroup functions based on their usage/aim/output similarity rather than technical differences. However, maybe we should rename/alias `savagedickey` to something more generic and less abstract... Like `bayesfactor_direct`, `bayesfactor_simple`, `bayesfactor_priorposterior`, `bayesfactor_ratio`...

> but so does bayesfactor_restrict, and it does get a separate function because the "what" the is being tested is different

I am not sure, could you please re-clarify the difference between the BF SD for a null interval and the BF restricted for the posterior outside vs. inside this interval?


Additional questions/thoughts:

- Would it make sense to add a mix between BF and ROPE (a BF based on the ROPE interval) to our comparison?
- `bayesfactor_restrict` or `bayesfactor_restricted`?

**Comment by mattansb** [comment]
I think what is in common to both the sdBF and the interval-null is that they test **univariate** prior and posterior distributions. (Does that give any inspiration for names?) They both also always give some BF<sub>*alternative vs. null*</sub> (or BF<sub>10</sub> for short). (But note that if the a directional test via `direction` it gives BF<sub>*direction-restricted vs. null*</sub>).

On the other hand, `bayesfactor_restrict` (probably `bayesfactor_restricted` is better) can be used to restrict **multivariate** prior and posterior distributions, and always gives a BF<sub>*restricted model vs. un-restricted model*</sub> (or BF<sub>r1</sub> for short). For example, the point-null says A=B=C, and the alternative (unrestricted) says A≠B≠C, but we can also test an order restricted model of A>B>C, or A>B & C>B, etc... So that BF<sub>r1</sub> tells us by how much more are the data probable under the restricted model vs. the unrestricted model. Richard Morey has a really good [blog post explaining all of this](https://richarddmorey.org/2015/01/multiple-comparisons-with-bayesfactor-part-2-order-restrictions/).

> Would it make sense to add a mix between BF and ROPE (a BF based on the ROPE interval) to our comparison?

In essence, the interval-null is exactly that! In practice, someone could pass the rope-range as the `null` argument!

## How are both types related?
Due to the transitive property of the BF, we can get from one to the other in some cases, like in the example I gave above. Let's name some models:
- M<sub>0</sub>: A-B=0
- M<sub>1</sub>: A-B≠0
- M<sub>2</sub>: 0.1<A-B<0.1 (in the plot above this is the area that is trapped between the dashed lines)
- M<sub>3</sub>: A-B < 0.1 OR 0.1 < A-B (the compliment to M<sub>2</sub>)

Thus, 
- BF<sub>10</sub> is the BF for the (unrestricted) alternative vs. the point null. (`bayesfactor_savagedickey(posterior,prior, null = 0)`)
- BF<sub>32</sub> is the BF for the (restricted) alternative vs the interval null (note that the alternative is restricted to not overlap with the interval-null model). (`bayesfactor_savagedickey(posterior,prior, null = c(-0.1, 0.1))`)
- BF<sub>21</sub> is the BF for the alternative (oddly specifying a small effect) vs. the (unrestricted) alternative . (`bayesfactor_restrict(posterior,prior, hypothesis = "-.1 < X & X < .1")`)
- BF<sub>31</sub> is the BF for the alternative (specifying a non-small effect) vs. the (unrestricted) alternative. (`bayesfactor_restrict(posterior,prior, hypothesis = "X > .1 | X < -.1")`)


Due to the transitive property of the BF, BF<sub>32</sub> = BF<sub>31</sub> / BF<sub>21</sub>!

(This is also exactly the point [Morey and Rouder (2011)](http://sci-hub.tw/http://doi.org/10.1037/a0024377) make in their paper... Our interval-null test is what they call there the "Nonoverlapping Hypotheses").

**Comment by strengejacke** [comment]
This looks really impressive! We should think about how we can keep the arguments as consistent as possible, so users can almost intuively switch between the different BF-methods.

Quick comments:

1) Yes, `bayesfactor_restricted()` sounds better
2) Do we really need `hypothesis = c("X > .1 | X < -.1","-.1 < X & X < .1")`, or can we simply call that argument `rope`, allowing a numeric vector of length 2? By that, we have automatically define the inside/outside-range.
3) If we need 2), but `rope` would also be possible, maybe we can have both, so users can't choose what they prefer?
4) We should probably add some details to the BF-vignettes, which explains the differences (both technically like ratio of prior/post. dist, not marg. lik. and practically, i.e. when to use which BF)
5) We need to modify the plotting-functions in **see**, or at least add another plot-method

**Comment by mattansb** [comment]
@strengejacke we commented at the same time, and I think I answered some of your questions there, but just so I make sure all you points have been addressed:

1. Agree.
2. `hypothesis = c("X > .1 | X < -.1","-.1 < X & X < .1")` was only one example I used to show how the restricted BF was similar to the interval-null BF. Normally this will not be the usage of the `bayesfactor_restricted`, it is less suitable for testing interval restrictions, and is more suitable for testing order restrictions (like what I gave in the first example).
3. The "rope BF" is the `bayesfactor_savagedickey` with the interval-null.
4. Absolutely. Once I get all the functions right I will update the vignette with all of this info.
5. Same as 4.

**Comment by strengejacke** [comment]
I'm not sure how much more we can add to such a plot, except maybe the shaded area for the null?

![image](https://user-images.githubusercontent.com/26301769/59973292-45354080-959e-11e9-9f9d-2943d6286701.png)


![image](https://user-images.githubusercontent.com/26301769/59973295-4d8d7b80-959e-11e9-84f0-7678fa1500e9.png)

**Comment by mattansb** [comment]
I thought adding an alpha aesthetic (but the shading works and is more
simple!) and removing the dots.

--
Mattan S. Ben-Shachar, PhD student
Department of Psychology & Zlotowski Center for Neuroscience
Ben-Gurion University of the Negev
The Developmental ERP Lab

**Comment by DominiqueMakowski** [comment]
> removing the dots.

:+1:

would it make sense to add the rope range as a default as null interval?

**Comment by mattansb** [comment]
> would it make sense to add the rope range as a default as null interval?

It would if the interval-null BF is it's own function - right now the default in the `bayesfactor_savagedickey` is the point null...

**Comment by DominiqueMakowski** [comment]
> It would if the interval-null BF is it's own function - right now the default in the bayesfactor_savagedickey is the point null...

right in that case let's keep it as is but maybe let put `'rope'` as argument to hypohesis in which cases it will automatically run rope range?

**Comment by mattansb** [comment]
How about a general function for univariate testing (`bayesfactor_simple` \ `bayesfactor_univariate` \ `bayesfactor_null` (**?**)), where:
- if `null` is a scalar it calls `bayesfactor_savagedickey` / `bayesfactor_nil` / `bayesfactor_pointnull` (**?**)
- if `null` is a range it calls `bayesfactor_nullinterval` / `bayesfactor_rope` / `bayesfactor_SESOI` / `bayesfactor_equivalence` (**?**) , that can have as a default `null = rope_range(posterior)`?

Any other name suggestions are welcome ^_^

**Comment by strengejacke** [comment]
I'm not sure I get the multivariate thing for the restricted BF. Do you have an example, including specifying the hypothesis, for a model with multiple variables?

**Comment by mattansb** [comment]
Sure. Here is the example from [Richard Morey's blog post](https://richarddmorey.org/2015/01/multiple-comparisons-with-bayesfactor-part-2-order-restrictions/):

Say you have 3 experimental condition, and you want a BF to see if they affect some DV. 
So you build two models and compare them:
```R
library(rstanarm)
library(bayestestR)
library(emmeans)

options(contrasts=c('contr.bayes', 'contr.bayes')) # from `bfrms` package

# replicating http://bayesfactor.blogspot.com/2015/01/multiple-comparisons-with-bayesfactor-2.html
disgust_data <- read.table(url('http://www.learnbayes.org/disgust_example.txt'),header=TRUE)

fit0 <- stan_glm(score ~ 1, data = disgust_data, family = gaussian(),
                 diagnostic_file = file.path(tempdir(), "df0.csv"),
                 prior = cauchy(0,0.5),
                 prior_aux = exponential(.1))
fit1 <- stan_glm(score ~ condition, data = disgust_data, family = gaussian(),
                 diagnostic_file = file.path(tempdir(), "df1.csv"),
                 prior = cauchy(0,0.5),
                 prior_aux = exponential(.1))
bayesfactor_models(fit0,fit1)
#> Computation of Bayes factors: estimating marginal likelihood, please wait...
#> Bayes factor analysis
#> ---------------------                     
#> [2] condition    0.26
#> 
#> Against denominator:
#> 		 [1] (Intercept only)   
#> ---
#> Bayes factor type:  marginal likelihoods (bridgesampling) 
```
We get a BF of 3.9 in favor of the null. But if you have some more specific hypothesis, such as that the conditions should be in some kind of order, and not just "different" from each other, the alternative tested above does not test that!

With order restrictions we can test several order restrictions at once (instead of testing A > B and then separately testing B > C - here we can test A > B > C in one BF!)
```R
em_condition <- emmeans(fit1, ~ condition)
hyps <- c("lemon < control & control < sulfur") #this is one hypothesis made of several order restrictions.

bayesfactor_restrict(em_condition, prior = fit1, hypothesis = hyps)
#> Computation of Bayes factors: sampling priors, please wait...
#> # Bayes Factor (Order-Restriction)
#> 
#>                          Hypothesis P(Prior) P(Posterior) Bayes Factor
#>  lemon < control & control < sulfur     0.19         0.74         3.99
#> ---
#> Bayes factors for the restricted movel vs. the un-restricted model.
```

Note that this BF is for the restricted model against the unrestricted model: BF<sub>r1</sub>, meaning the data are 3.99 more probable under a model with the order restriction than with the model without any restriction!
If we want BF<sub>r0</sub>, we can get it with BF<sub>r1</sub> * BF<sub>10</sub> = 3.99 * 0.26 = 1.03.
Thus the data is equally probable under a model with no effect at all for condition *or* a model with a specific order restriction.

In other words, if you have a hypothesis say that the effect of X is positive, and that the difference between A1 and A2 is negative, you can get a single BF representing a single model that has these (order) restrictions...

**Comment by DominiqueMakowski** [comment]
naive question: is it strongly different from a "regular" bf against the bull based on the marginal contrasts (e. g. obtained via emmeans)?

**Comment by mattansb** [comment]
On the marginal means, you would get 2 BFs with different alt and nulls:
- First BF: null is no diff between lemon and control, alt is some diff between lemon and control (all other parameters / diffs are assumed to be non-null).
- First BF: null is no diff between sulfur and control, alt is some diff between sulfur and control (all other parameters / diffs are assumed to be non-null).

I am not aware of any way to incorporate these two BFs into 1...

**Comment by strengejacke** [comment]
> instead of testing A > B and then separately testing B > C - here we can test A > B > C in one BF!

I think this is clear. So this is again something slightly different than testing if an effects is outside a ROPE, say `A < -.5 & A > .5`?

> Thus the data is equally probable under a model with no effect at all for condition or a model with a specific order restriction.

I don't quite get this. Isn't a model _without any restriction_ the same as _ no effect at all for condition_, so here we have either BF = 3.99  or BF= 1 / 3.99 ?

**Comment by mattansb** [comment]
A model with *no effect* is a model **with** a restriction, the restriction of: *theta = 0*!

An unrestricted model is the full model, where the parameters are free to be any value. In such an unrestricted model, we can have any of the following orders (though some are more probable than other *a posteriori*, they are all equally probable *a priori*):
A > B > C,
A > C > B,
B > A >C,
B > C > A, 
C > A > B,
C > B > A.

`bayesfactor_restricted` computes the marginal likelihood of an order-restricted model, in which only some orders of the parameters are "allowed" *a priori*, and compares it the marginal likelihood of the unrestricted model. So you get BF<sub>r1</sub> and not BF<sub>r0</sub> (which you can get by: BF<sub>r0</sub> = BF<sub>r1</sub> / BF<sub>10</sub>, but that will over complicate the function, and we have a function the does BF<sub>10</sub> - `bayesfactor_models`).

In theory you can also test interval restrictions with `bayesfactor_restricted`, but because of the way `bayesfactor_restricted` works, and the fact that it returns BF<sub>r1</sub>, it is better to use the interval-null option in the sdBF function (where these interval restrictions are testing with a logspline probability function, while `bayesfactor_restricted` just "counts" how many samples from the prior and posterior draws meet the order restriction).

### Summary - model types
- M<sub>1</sub>: **Unrestricted model** - *a priori* X can have any effect, with some more likely than others (specified by a prior density distribution). This is the model you would normally get your estimates from if you're not into Bayes factors 😅.
- M<sub>2</sub>: **Point model** - *a priori* the effect of X is exactly some value.
- M<sub>3</sub>: **Interval model** - *a priori* the effect of X is restricted to a some interval.
- M<sub>4</sub>: **Order restricted model** - *a priori* the effect of X is restricted so that only specific orders between the parameters are allowed.

### BFs types (and the model types they correspond to)
- `bayesfactor_savagedickey` / `bayesfactor_models` give a BF of M<sub>1</sub> vs. M<sub>2</sub> (the point being 0).
- `bayesfactor_savagedickey` when `null` is a range, gives a BF of a M<sub>3</sub> (restricted to *outside* the null-interval range) vs. M<sub>3</sub> (restricted to *inside* the null-interval range).
- `bayesfactor_savagedickey` when `direction` is set, gives a BF of a M<sub>3</sub> (restricted to positive range) vs. M<sub>2</sub> (the point being 0).
- `bayesfactor_savagedickey` when `null` is a range **&** `direction` is set, gives a BF of M<sub>3</sub> (restricted to positive range outside the null-interval) vs. M<sub>3</sub> (restricted to *inside* the null-interval range).
- `bayesfactor_restricted` gives a BF of some M<sub>4</sub> vs. M<sub>1</sub>.

I hope that's more clear? This is definitely the hardest Bayes factor to grasp... I had to read the blog post and both papers several times, and I only really got the gist when I started writing `BFEffects::restrict` some time ago...

**Comment by strengejacke** [comment]
> I thought adding an alpha aesthetic

```r
Error in f(...) : Aesthetics can not vary with a ribbon
```

**Comment by mattansb** [comment]
That settles that then 😅

**Comment by strengejacke** [comment]
I think what we can add to the docs is a one-sentence summary for all BF methods we have. Correct me if I'm wrong, but what I understood is:

- BF models compares models against a denominator model (saying how much better/worse a model performs against the denominator model)
- BF inclusion compares models against each other (saying which model performs best)
- BF SD compares model parameters against the NULL ("p-values")
- BF restricted compares if a model where parameters are assumed to have stronger/weaker effects than other parameters performs better/worse than a model where we make no assumption about the strength of effects among parameters

**Comment by mattansb** [comment]
- BF inclusion generates evidence for specific effects by Bayesian model averaging (comparing in some way all models to each other, by effect).

You would want this in cases where you're not really interested in a specific model, but you are interested in seeing how effect X does across several models. Personally I don't really use this, but it is a popular feature of JASP, so...

**Comment by strengejacke** [comment]
Does it make sense to have a plot-method for BF restricted? I'm currently not sure how this could look like.

**Comment by mattansb** [comment]
Don't think so... at best it can be passed to the BF-models' pizza plot? Meh..

**Comment by DominiqueMakowski** [comment]
Great work 🥇 

I fear like the challenge now will become to show the users how/when these different versions can be relevant for their questions. Which means robust documentation and in the long run, posts demonstrating some use cases. But I don't worry about it at all, it'll come with time. As we ourselves become familiar with these tools/indices and we start using them for our questions, it will help us deliver interesting and useful examples ☺️ 

> Don't think so... at best it can be passed to the BF-models' pizza plot? Meh.. 

🍕👍 

- Is it possible to add BF-restricted to the general `bayesfactor` function, so it is smart enough to know that if the user passes a hypothesis of a specific form (for instance containing parameters of the model) if it will run it? 
- Kind of related to the above, I am also not quite sure about the names and all, since the functions expanded much (and well). But it's indeed tough to find a names-framework that is intuitive and correct. It's on my mind though

**Comment by mattansb** [comment]
> Is it possible to add BF-restricted to the general `bayesfactor` function...

I think so. I've added this to my to-do list (above).

The various docs will need to be heavily expanded indeed... I think it is best to let this PR cook slowly, until the functions are well documented (also with proper names), with explanations and examples (which are a major feature of our ❤📦).

Generally I don't think we're in a hurry with these two BFs - the major BF functions have already been implemented (well), are well documented, and these new BFs are very (very) niche.

**Comment by strengejacke** [comment]
True! but...

> and these new BFs are very (very) niche.

... and we can be one of the firsts to implement it! :-)

**Comment by strengejacke** [comment]
Over the course of the past weeks, I started liking bayes factors. One reason is for sure the nice implementation in bayestestR and see, but - apart from critism especially from Bayesian statisticians - I think these are reasonable additions to the classical NHST in frequentist frameworks. As far as I am concerned, I'm not completely devoted to either the Bayesian or Frequentist side of statistics, so having all these tools (also in our other packages) is really great!

**Comment by DominiqueMakowski** [comment]
>  I think it is best to let this PR cook slowly

 Beware, Daniel's blood will be boiling seeing this unmerged PR hanging in there 😅 

> Over the course of the past weeks, I started liking bayes factors.

a big hats off to @mattansb for successfully shifting @strengejacke's posteriors :)

But, I tend to agree :) however, I am still very curious regarding these impressive precepts in terms of sampling (e.g., "at least 10000 draws to have accurate BFs"). I am still wondering whether the accuracy for regular models (e.g. rstanarm's 4000 default) is *this much* off or is it, in the end, quite usable...

**Comment by strengejacke** [comment]
> I am still wondering whether the accuracy for regular models (e.g. rstanarm's 4000 default) is this much off or is it, in the end, quite usable...

Looks like another paper... There's much content we can generate from this package ;-)

**Comment by mattansb** [comment]
> ... and we can be one of the firsts to implement it! :-)

Actually... the order restrictions have recently been implemented in JASP (but I didn't know this when I started working on our function), and the null interval can, with some work, be implemented in brms... but *obviously* ours is better!

> Over the course of the past weeks, I started liking bayes factors.

My work here is done! 😜

>  "at least 10000 draws to have accurate BFs"

Nope, this was a mistake on my part - 4000 seems to be fine for normal usage. Lets say you need 200 draws per order restriction (for a 0.5% precision) (one order restriction: A >B, two: A > B & C > D), that mean you could have up to 200 (orthogonal!) order restrictions, which is likely to never ever happen (BF = 7*10^42). I will amend this in the doc ^_^

**Comment by mattansb** [comment]
I think after the functions here as solid, I will go one to .. basically re-write (or at least re-structure) the BF vignette into the following structure (putting this here for some feedback & so I don't forget):

1. **General Intro to BFs** (framing BFs as any change from prior odds to posterior odds, not just as likelihood-ratios).
2. **Univariate BFs**
2.1. *Null-Interval BFs* (+ "one sided tests")
2.2. *sdBF* - Show how it is equal to the null-interval as the interval shrinks to a point.
3. **Comparing models**
3.1. *marginal likelihood* (=evidence) and more (BIC approx.)
3.2. *model averaging* (inclusion BFs)
3.3. *order restricted models*

(Still holding out for function names. `bayesfactor_univariate`? `bayesfactor_parameter` [vs. `bayesfactor_models`]?)

**Comment by mattansb** [comment]
I think after the functions here as solid, I will go one to .. basically re-write (or at least re-structure) the BF vignette into the following structure (putting this here for some feedback & so I don't forget):

1. **General Intro to BFs** (framing BFs as any change from prior odds to posterior odds, not just as likelihood-ratios).
2. **Univariate BFs**
2.1. *Null-Interval BFs* (+ "one sided tests")
2.2. *sdBF* - Show how it is equal to the null-interval as the interval shrinks to a point.
3. **Comparing models**
3.1. *marginal likelihood* (=evidence) and more (BIC approx.)
3.2. *model averaging* (inclusion BFs)
3.3. *order restricted models*

(Still holding out for function names. `bayesfactor_univariate`? `bayesfactor_parameter` [vs. `bayesfactor_models`]?)

**Comment by strengejacke** [comment]
Sounds good! :-)

**Comment by DominiqueMakowski** [comment]
`bayesfactor_parameters` vs. `bayesfactor_models` is nice and simple!

Although bayesfactor inclusion are also about *parameters* (giving the "average" BF for each parameter), and yet the way it works is closer to bayesfactor_models...

**Comment by mattansb** [comment]
BF inclusion tests *terms* - so a factor will get a single inclusion BF even if has, in the model, 10 parameters. I will make sure this is clear in the (updated) vignette.

Ok, so we'll go with `bayesfactor_parameters`:
- When `length(null)==1` will pass to `bayesfactor_savagedickey` (default)
- When `length(null)==2` will pass to... ?.... `bayesfactor_interval`? **`bayesfactor_rope`**??

**Comment by mattansb** [comment]
I'm closing this as all major points have been addressed. Finer points can be discussed at #194
